runtime/common/ArgumentConversion.cpp

```diff
@@ -100,6 +92,60 @@ static Value genConstant(OpBuilder &, cudaq::cc::StructType, void *,
 static Value genConstant(OpBuilder &, cudaq::cc::ArrayType, void *,
                          ModuleOp substMod, llvm::DataLayout &);
 
+static Value genConstant(OpBuilder &builder, const cudaq::state *v,
+                         ModuleOp substMod, llvm::DataLayout &layout,
+                         bool isSimulator) {
+  if (isSimulator) {
+    // The program is executed remotely, materialize the simulation data
+    // into an array and create a new state from it.
+    auto ctx = builder.getContext();
+    auto loc = builder.getUnknownLoc();
+    auto size = 1ULL << v->get_num_qubits();
+    auto is64Bit =
+        v->get_precision() == cudaq::SimulationState::precision::fp64;
+    auto eleTy = is64Bit ? ComplexType::get(Float64Type::get(ctx))
+                         : ComplexType::get(Float32Type::get(ctx));
+    auto arrTy = cudaq::cc::ArrayType::get(ctx, eleTy, size);
+
+    auto genConArray = [&]<typename T>() -> Value {
+      std::complex<T> data[size];
+      for (std::size_t i = 0; i < size; i++) {
+        data[i] = (*v)({i}, 0);
```

Comment:
@1tnguyen accessing the state elements here (using the const functions from state.cpp).